File: backend/schemas/game.py
"""Game-related schemas for TRPG (ClaudeWorld)."""

from datetime import datetime
from typing import Any, List, Optional

from domain.services.player_state_serializer import PlayerStateSerializer
from domain.value_objects.enums import Language, WorldPhase
from i18n.serializers import serialize_utc_datetime as _serialize_utc_datetime
from pydantic import BaseModel, field_serializer, model_validator

# =============================================================================
# World Schemas
# =============================================================================


class WorldBase(BaseModel):
    """Base schema for world data."""

    name: str
    user_name: Optional[str] = None  # Player's display name (set during onboarding, defaults to 여행자/traveler)
    language: Language = Language.ENGLISH  # UI/message language


class WorldCreate(WorldBase):
    """Schema for creating a new world."""

    pass


class WorldUpdate(BaseModel):
    """Schema for updating a world."""

    phase: Optional[WorldPhase] = None
    genre: Optional[str] = None
    theme: Optional[str] = None
    user_name: Optional[str] = None
    stat_definitions: Optional[dict] = None


class StatDefinition(BaseModel):
    """Schema for a single stat definition."""

    name: str
    display: str
    min: Optional[int] = None
    max: Optional[int] = None
    default: int = 0
    color: Optional[str] = None


class StatDefinitions(BaseModel):
    """Schema for stat definitions wrapper."""

    stats: List[StatDefinition] = []


class WorldSummary(WorldBase):
    """Schema for world listing."""

    id: int
    owner_id: Optional[str] = None
    phase: WorldPhase = WorldPhase.ONBOARDING
    genre: Optional[str] = None
    theme: Optional[str] = None
    onboarding_room_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    last_played_at: Optional[datetime] = None

    @field_serializer("created_at")
    def serialize_created_at(self, dt: datetime, _info):
        return _serialize_utc_datetime(dt)

    @field_serializer("updated_at")
    def serialize_updated_at(self, dt: datetime, _info):
        return _serialize_utc_datetime(dt)

    @field_serializer("last_played_at")
    def serialize_last_played_at(self, dt: Optional[datetime], _info):
        return _serialize_utc_datetime(dt) if dt else None

    class Config:
        from_attributes = True


class World(WorldSummary):
    """Full world schema with all details."""

    stat_definitions: Optional[StatDefinitions] = None
    lore: Optional[str] = None  # Loaded from filesystem

    @model_validator(mode="before")
    @classmethod
    def parse_stat_definitions(cls, data: Any) -> Any:
        """Parse stat_definitions from JSON string if stored."""
        if hasattr(data, "__dict__"):
            stat_defs = getattr(data, "stat_definitions", None)
            if stat_defs and isinstance(stat_defs, str):
                import json

                try:
                    parsed = json.loads(stat_defs)
                    # Return nested structure { stats: [...] } for frontend compatibility
                    stats_data = {"stats": parsed.get("stats", [])} if isinstance(parsed, dict) else {"stats": parsed}
                    data_dict = {
                        "id": data.id,
                        "name": data.name,
                        "user_name": data.user_name,
                        "language": data.language,
                        "owner_id": data.owner_id,
                        "phase": data.phase,
                        "genre": data.genre,
                        "theme": data.theme,
                        "onboarding_room_id": data.onboarding_room_id,
                        "stat_definitions": stats_data,
                        "created_at": data.created_at,
                        "updated_at": data.updated_at,
                        "last_played_at": data.last_played_at,
                    }
                    return data_dict
                except (json.JSONDecodeError, TypeError):
                    pass
        return data

    class Config:
        from_attributes = True


class ImportableWorld(BaseModel):
    """Schema for a world that exists in filesystem but not in database."""

    name: str
    owner_id: Optional[str] = None
    user_name: Optional[str] = None
    language: Language = Language.ENGLISH
    phase: WorldPhase = WorldPhase.ONBOARDING
    genre: Optional[str] = None
    theme: Optional[str] = None
    created_at: Optional[datetime] = None

    @field_serializer("created_at")
    def serialize_created_at(self, dt: Optional[datetime], _info):
        return _serialize_utc_datetime(dt) if dt else None


class WorldResetRequest(BaseModel):
    """Request schema for resetting a world to its initial state."""

    confirm: bool = False  # Safety flag to confirm reset


class WorldResetResponse(BaseModel):
    """Response schema after world reset."""

    success: bool
    message: str
    world_id: int
    starting_location: str


# =============================================================================
# Location Schemas
# =============================================================================


class LocationBase(BaseModel):
    """Base schema for location data."""

    name: str
    display_name: Optional[str] = None
    description: Optional[str] = None


class LocationCreate(LocationBase):
    """Schema for creating a new location."""

    position_x: int = 0
    position_y: int = 0
    adjacent_to: Optional[List[int]] = None
    is_discovered: bool = True


class LocationUpdate(BaseModel):
    """Schema for updating a location."""

    label: Optional[str] = None
    is_discovered: Optional[bool] = None


class Location(LocationBase):
    """Full location schema."""

    id: int
    world_id: int
    label: Optional[str] = None
    position_x: int = 0
    position_y: int = 0
    adjacent_locations: Optional[List[int]] = None
    room_id: Optional[int] = None
    is_current: bool = False
    is_discovered: bool = True

    @model_validator(mode="before")
    @classmethod
    def parse_adjacent_locations(cls, data: Any) -> Any:
        """Parse adjacent_locations from JSON string if stored."""
        if hasattr(data, "__dict__"):
            adjacent = getattr(data, "adjacent_locations", None)
            if adjacent and isinstance(adjacent, str):
                import json

                try:
                    parsed = json.loads(adjacent)
                    data_dict = {
                        "id": data.id,
                        "world_id": data.world_id,
                        "name": data.name,
                        "display_name": data.display_name,
                        "description": data.description,
                        "label": data.label,
                        "position_x": data.position_x,
                        "position_y": data.position_y,
                        "adjacent_locations": parsed,
                        "room_id": data.room_id,
                        "is_current": data.is_current,
                        "is_discovered": data.is_discovered,
                    }
                    return data_dict
                except (json.JSONDecodeError, TypeError):
                    pass
        return data

    class Config:
        from_attributes = True


# =============================================================================
# Player State Schemas
# =============================================================================


class GameTime(BaseModel):
    """Schema for in-game time."""

    hour: int = 8
    minute: int = 0
    day: int = 1


class InventoryItem(BaseModel):
    """Schema for an inventory item."""

    id: str
    name: str
    description: Optional[str] = None
    quantity: int = 1
    properties: Optional[dict] = None


class PlayerStateBase(BaseModel):
    """Base schema for player state."""

    turn_count: int = 0


class PlayerState(PlayerStateBase):
    """Full player state schema."""

    id: int
    world_id: int
    current_location_id: Optional[int] = None
    current_location_name: Optional[str] = None
    stats: Optional[dict] = None
    inventory: Optional[List[InventoryItem]] = None
    effects: Optional[List[dict]] = None
    action_history: Optional[List[dict]] = None
    # Chat mode state
    is_chat_mode: bool = False
    chat_mode_start_message_id: Optional[int] = None
    # In-game time
    game_time: Optional[GameTime] = None

    @model_validator(mode="before")
    @classmethod
    def parse_json_fields(cls, data: Any) -> Any:
        """Parse JSON fields from database storage."""
        if hasattr(data, "__dict__"):
            data_dict = {
                "id": data.id,
                "world_id": data.world_id,
                "current_location_id": data.current_location_id,
                "turn_count": data.turn_count,
            }

            # Parse stats using PlayerStateSerializer
            stats = getattr(data, "stats", None)
            data_dict["stats"] = PlayerStateSerializer.parse_stats(stats)

            # Parse inventory using PlayerStateSerializer and normalize item_id -> id
            inventory = getattr(data, "inventory", None)
            parsed_inventory = PlayerStateSerializer.parse_inventory(inventory)
            # Normalize: ensure each item has 'id' (filesystem uses 'item_id')
            if parsed_inventory:
                for item in parsed_inventory:
                    if "item_id" in item and "id" not in item:
                        item["id"] = item.pop("item_id")
            data_dict["inventory"] = parsed_inventory

            # Parse effects using PlayerStateSerializer
            effects = getattr(data, "effects", None)
            data_dict["effects"] = PlayerStateSerializer.parse_effects(effects)

            # Parse action_history using PlayerStateSerializer
            action_history = getattr(data, "action_history", None)
            data_dict["action_history"] = PlayerStateSerializer.parse_action_history(action_history)

            # Get current location name if relationship exists
            current_location = getattr(data, "current_location", None)
            if current_location:
                data_dict["current_location_name"] = current_location.display_name or current_location.name

            # Chat mode fields
            data_dict["is_chat_mode"] = getattr(data, "is_chat_mode", False) or False
            data_dict["chat_mode_start_message_id"] = getattr(data, "chat_mode_start_message_id", None)

            # Game time (from filesystem, may not exist in DB model)
            game_time = getattr(data, "game_time", None)
            if game_time:
                data_dict["game_time"] = game_time

            return data_dict
        return data

    class Config:
        from_attributes = True


class PlayerAction(BaseModel):
    """Schema for a player action."""

    text: str
    image_data: Optional[str] = None  # Base64 encoded image data
    image_media_type: Optional[str] = None  # MIME type (e.g., 'image/png')


class GameStateResponse(BaseModel):
    """Schema for the full game state response."""

    world: WorldSummary
    player_state: PlayerState
    current_location: Optional[Location] = None
    suggestions: Optional[List[str]] = None


__all__ = [
    # World
    "WorldBase",
    "WorldCreate",
    "WorldUpdate",
    "StatDefinition",
    "StatDefinitions",
    "WorldSummary",
    "World",
    "ImportableWorld",
    "WorldResetRequest",
    "WorldResetResponse",
    # Location
    "LocationBase",
    "LocationCreate",
    "LocationUpdate",
    "Location",
    # Player State
    "GameTime",
    "InventoryItem",
    "PlayerStateBase",
    "PlayerState",
    "PlayerAction",
    "GameStateResponse",
]

File: backend/src/schemas/game.ts
// Three models are not plain reflections of a table: `World` merges a row with
// `lore.md` and `stats.yaml`, `PlayerState` merges one with the resolved
// inventory/clock/equipment from `player.yaml`, and `Location` reproduces a live
// bug — see `toLocation`.

import { z } from 'zod'
import type { PlayerStateWithLocation } from '../crud/player-state'
import type { Location as LocationRow, World as WorldRow } from '../db/schema'
import { LANGUAGES, WORLD_PHASES } from '../db/schema'
import type { InventoryEntry } from '../domain/player-rules'
import {
  isoDatetime,
  optionalBool,
  optionalInt,
  optionalString,
  parseJsonColumn,
  pydanticBool,
  pydanticInt,
  requiredTimestamp,
  serializeBool,
  serializeOptionalUtcDatetime,
} from './common'

export const WorldBase = z.object({
  name: z.string(),
  /** The player's display name *inside* the world, set during onboarding. */
  user_name: optionalString(),
  language: z.enum(LANGUAGES).default('en'),
})

export type WorldBase = z.infer<typeof WorldBase>

export const WorldCreate = WorldBase

export type WorldCreate = z.infer<typeof WorldCreate>

/** `stat_definitions` is deliberately open: the update path writes it into the
 * TEXT column as-is, while the read path wraps it in `{stats: []}`. */
export const WorldUpdate = z.object({
  phase: z.enum(WORLD_PHASES).nullable().default(null),
  genre: optionalString(),
  theme: optionalString(),
  user_name: optionalString(),
  stat_definitions: z.record(z.string(), z.unknown()).nullable().default(null),
})

export type WorldUpdate = z.infer<typeof WorldUpdate>

/** As authored in `worlds/<world>/stats.yaml`. `max` is optional because uncapped
 * stats exist; `min`/`color` have UI fallbacks. */
export const StatDefinition = z.object({
  name: z.string(),
  display: z.string(),
  min: optionalInt(),
  max: optionalInt(),
  default: pydanticInt().default(0),
  color: optionalString(),
})

export type StatDefinition = z.infer<typeof StatDefinition>

/** `GameStatePanel` reads `world.stat_definitions?.stats`, so the array cannot be
 * hoisted even though nothing else uses the wrapper. */
export const StatDefinitions = z.object({
  stats: StatDefinition.array().default([]),
})

export type StatDefinitions = z.infer<typeof StatDefinitions>

export const WorldSummary = WorldBase.extend({
  id: pydanticInt(),
  owner_id: optionalString(),
  phase: z.enum(WORLD_PHASES).default('onboarding'),
  genre: optionalString(),
  theme: optionalString(),
  onboarding_room_id: optionalInt(),
  created_at: isoDatetime(),
  updated_at: isoDatetime(),
  last_played_at: isoDatetime().nullable().default(null),
})

export type WorldSummary = z.infer<typeof WorldSummary>

export const World = WorldSummary.extend({
  stat_definitions: StatDefinitions.nullable().default(null),
  lore: optionalString(),
})

export type World = z.infer<typeof World>

/** A world in `worlds/` but absent from the database — offered for import. Built
 * from a `WorldConfig`, never a row, so it has no `id` and a hand-made world
 * folder need not carry a `created_at`. */
export const ImportableWorld = z.object({
  name: z.string(),
  owner_id: optionalString(),
  user_name: optionalString(),
  language: z.enum(LANGUAGES).default('en'),
  phase: z.enum(WORLD_PHASES).default('onboarding'),
  genre: optionalString(),
  theme: optionalString(),
  created_at: isoDatetime().nullable().default(null),
})

export type ImportableWorld = z.infer<typeof ImportableWorld>

/** `confirm` is a deliberate speed bump: a reset wipes the world's progress. */
export const WorldResetRequest = z.object({
  confirm: pydanticBool().default(false),
})

export type WorldResetRequest = z.infer<typeof WorldResetRequest>

export const WorldResetResponse = z.object({
  success: z.boolean(),
  message: z.string(),
  world_id: pydanticInt(),
  /** Display name of the location the player was returned to. */
  starting_location: z.string(),
})

export type WorldResetResponse = z.infer<typeof WorldResetResponse>

export function toWorldSummary(row: WorldRow): WorldSummary {
  return {
    name: row.name,
    user_name: row.userName,
    language: row.language ?? 'en',
    id: row.id,
    owner_id: row.ownerId,
    phase: row.phase ?? 'onboarding',
    genre: row.genre,
    theme: row.theme,
    onboarding_room_id: row.onboardingRoomId,
    created_at: requiredTimestamp(row.createdAt, 'WorldSummary', 'created_at'),
    updated_at: requiredTimestamp(row.updatedAt, 'WorldSummary', 'updated_at'),
    last_played_at: serializeOptionalUtcDatetime(row.lastPlayedAt),
  }
}

/** Both fields *replace* rather than merge: the `worlds.stat_definitions` column
 * is a cache the read path ignores. */
export interface WorldFilesystemOverlay {
  lore: string | null
  /** Already in `{stats: [...]}` form. */
  stat_definitions: StatDefinitions | null
}

/** Without an overlay `lore` is null and `stat_definitions` comes from the TEXT
 * column, whose shape is not fixed — `{"stats": []}` and a bare `[]` are both in
 * the wild. Every real caller passes the overlay. */
export function toWorld(row: WorldRow, overlay?: WorldFilesystemOverlay): World {
  return {
    ...toWorldSummary(row),
    stat_definitions: overlay ? overlay.stat_definitions : parseStatDefinitionsColumn(row.statDefinitions),
    lore: overlay ? overlay.lore : null,
  }
}

/** Null when the column is unreadable. */
export function parseStatDefinitionsColumn(raw: string | null): StatDefinitions | null {
  return parseJsonColumn(
    raw,
    z.union([StatDefinitions, StatDefinition.array().transform((stats) => ({ stats }))]),
  )
}

/** **Throws** on a malformed entry, unlike the DB-blob decoders that fall back to
 * null: a hand-authored file breaking should be loud. */
export function toStatDefinitions(raw: readonly Record<string, unknown>[]): StatDefinitions {
  return { stats: raw.map((entry) => StatDefinition.parse(entry)) }
}

/** Typed structurally so this module does not depend on the service layer.
 * `language`/`phase` are unvalidated on the config, so both are narrowed here. */
export function toImportableWorld(config: {
  name: string
  ownerId: string | null
  userName: string | null
  language: string
  phase: string
  genre: string | null
  theme: string | null
  createdAt: Date | null
}): ImportableWorld {
  return {
    name: config.name,
    owner_id: config.ownerId,
    user_name: config.userName,
    language: oneOf(LANGUAGES, config.language, 'en'),
    phase: oneOf(WORLD_PHASES, config.phase, 'onboarding'),
    genre: config.genre,
    theme: config.theme,
    created_at: serializeOptionalUtcDatetime(config.createdAt),
  }
}

function oneOf<T extends string>(allowed: readonly T[], value: string, fallback: T): T {
  return (allowed as readonly string[]).includes(value) ? (value as T) : fallback
}

export const LocationBase = z.object({
  /** The folder name under `worlds/<world>/locations/`, e.g. `old_mill`. */
  name: z.string(),
  display_name: optionalString(),
  description: optionalString(),
})

export type LocationBase = z.infer<typeof LocationBase>

export const LocationCreate = LocationBase.extend({
  position_x: pydanticInt().default(0),
  position_y: pydanticInt().default(0),
  /** Named `adjacent_to` in, `adjacent_locations` out; the column is the latter. */
  adjacent_to: pydanticInt().array().nullable().default(null),
  is_discovered: pydanticBool().default(true),
  /** Set while the Location Designer sub-agent has yet to enrich the stub. */
  is_draft: pydanticBool().default(false),
})

export type LocationCreate = z.infer<typeof LocationCreate>

export const LocationUpdate = z.object({
  name: optionalString(),
  display_name: optionalString(),
  description: optionalString(),
  /** The one field the *user* can set; `PATCH .../locations/{id}` reads only this. */
  label: optionalString(),
  position_x: optionalInt(),
  position_y: optionalInt(),
  is_discovered: optionalBool(),
  is_draft: optionalBool(),
})

export type LocationUpdate = z.infer<typeof LocationUpdate>

export const Location = LocationBase.extend({
  id: pydanticInt(),
  world_id: pydanticInt(),
  label: optionalString(),
  position_x: pydanticInt().default(0),
  position_y: pydanticInt().default(0),
  adjacent_locations: pydanticInt().array().nullable().default(null),
  room_id: optionalInt(),
  is_current: pydanticBool().default(false),
  is_discovered: pydanticBool().default(true),
  is_draft: pydanticBool().default(false),
})

export type Location = z.infer<typeof Location>

/** **Reproduces a live bug, deliberately:** when `adjacent_locations` holds JSON,
 * `is_draft` reads `false` whatever the column says. Nothing in `frontend/` reads
 * it. */
export function toLocation(row: LocationRow): Location {
  const adjacent = row.adjacentLocations
    ? parseJsonColumn(row.adjacentLocations, pydanticInt().array())
    : null
  const rebuilt = adjacent !== null

  return {
    name: row.name,
    display_name: row.displayName,
    description: row.description,
    id: row.id,
    world_id: row.worldId,
    label: row.label,
    position_x: row.positionX ?? 0,
    position_y: row.positionY ?? 0,
    adjacent_locations: adjacent,
    room_id: row.roomId,
    is_current: serializeBool(row.isCurrent),
    is_discovered: row.isDiscovered ?? true,
    is_draft: rebuilt ? false : serializeBool(row.isDraft),
  }
}

/** The defaults are hour 8 of day 1 — a world opens at dawn. */
export const GameTime = z.object({
  hour: pydanticInt().default(8),
  minute: pydanticInt().default(0),
  day: pydanticInt().default(1),
})

export type GameTime = z.infer<typeof GameTime>

/** `player.yaml` keys the reference `item_id` and the response calls it `id`;
 * {@link toInventoryItem} renames it so nothing downstream knows both. */
export const InventoryItem = z.object({
  id: z.string(),
  name: z.string(),
  description: optionalString(),
  quantity: pydanticInt().default(1),
  properties: z.record(z.string(), z.unknown()).nullable().default(null),
})

export type InventoryItem = z.infer<typeof InventoryItem>

export const PlayerStateBase = z.object({
  turn_count: pydanticInt().default(0),
})

export type PlayerStateBase = z.infer<typeof PlayerStateBase>

export const PlayerState = PlayerStateBase.extend({
  id: pydanticInt(),
  world_id: pydanticInt(),
  current_location_id: optionalInt(),
  /** Denormalized from the joined location for the header; not a column. */
  current_location_name: optionalString(),
  stats: z.record(z.string(), z.unknown()).nullable().default(null),
  inventory: InventoryItem.array().nullable().default(null),
  effects: z.record(z.string(), z.unknown()).array().nullable().default(null),
  action_history: z.record(z.string(), z.unknown()).array().nullable().default(null),
  is_chat_mode: pydanticBool().default(false),
  /** Where the transcript resumes when chat mode is left. */
  chat_mode_start_message_id: optionalInt(),
  game_time: GameTime.nullable().default(null),
  /** `slot_name -> item_id`, with `null` for an empty slot. */
  equipment: z.record(z.string(), z.unknown()).nullable().default(null),
})

export type PlayerState = z.infer<typeof PlayerState>

export const PlayerAction = z.object({
  text: z.string(),
  image_data: optionalString(),
  image_media_type: optionalString(),
})

export type PlayerAction = z.infer<typeof PlayerAction>

/** Unused: no route returns it — `GET .../state` returns a bare `PlayerState`. */
export const GameStateResponse = z.object({
  world: WorldSummary,
  player_state: PlayerState,
  current_location: Location.nullable().default(null),
  suggestions: z.string().array().nullable().default(null),
})

export type GameStateResponse = z.infer<typeof GameStateResponse>

/** The `player.yaml`-sourced half of a player-state response; the matching
 * columns are a cache the read path does not consult. */
export interface PlayerStateOverlay {
  inventory: InventoryEntry[]
  gameTime: GameTime | null
  equipment: Record<string, string | null> | null
}

/** `stats`, `effects` and `action_history` come back `null` when their column is
 * empty, *not* `{}` / `[]`. */
export function toPlayerState(row: PlayerStateWithLocation, overlay: PlayerStateOverlay): PlayerState {
  const location = row.currentLocation
  return {
    turn_count: row.turnCount ?? 0,
    id: row.id,
    world_id: row.worldId,
    current_location_id: row.currentLocationId,
    current_location_name: location ? (location.displayName ?? location.name) : null,
    stats: parseJsonColumn(row.stats, z.record(z.string(), z.unknown())),
    inventory: overlay.inventory.map(toInventoryItem),
    effects: parseJsonColumn(row.effects, z.record(z.string(), z.unknown()).array()),
    action_history: parseJsonColumn(row.actionHistory, z.record(z.string(), z.unknown()).array()),
    is_chat_mode: serializeBool(row.isChatMode),
    chat_mode_start_message_id: row.chatModeStartMessageId,
    game_time: overlay.gameTime,
    equipment: overlay.equipment,
  }
}

/** `item_id` wins over `id`, and an entry with neither — or with no name —
 * becomes `""` rather than an error: a half-written `player.yaml` produces those
 * shapes mid-turn, and a blank row beats 500-ing the panel. */
export function toInventoryItem(entry: InventoryEntry): InventoryItem {
  return {
    id: entry.item_id || entry.id || '',
    name: entry.name ?? '',
    description: entry.description ?? null,
    quantity: entry.quantity ?? 1,
    properties: entry.properties ?? null,
  }
}
